File: src/main/java/pw/mihou/nexus/features/command/interceptors/commons/modules/ratelimiter/core/NexusRatelimiterCore.java
package pw.mihou.nexus.features.command.interceptors.commons.modules.ratelimiter.core;

import pw.mihou.nexus.core.NexusCore;
import pw.mihou.nexus.features.command.core.NexusCommandCore;
import pw.mihou.nexus.features.command.facade.NexusCommand;
import pw.mihou.nexus.features.command.facade.NexusMiddlewareEvent;
import pw.mihou.nexus.features.command.interceptors.commons.modules.ratelimiter.facade.NexusRatelimitData;
import pw.mihou.nexus.features.command.interceptors.commons.modules.ratelimiter.facade.NexusRatelimiter;

import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;

public class NexusRatelimiterCore implements NexusRatelimiter {

    private final Map<Entity, NexusRatelimitData> ratelimits = new ConcurrentHashMap<>();

    /**
     * Ratelimits a user on a specific server, or on their private channel for a
     * specific command.
     *
     * @param user      The user to rate-limit.
     * @param server    The server or private channel to rate-limit on.
     * @param command   The command to rate-limit on.
     * @return          The results from the rate-limit attempt.
     */
    private AccessorRatelimitData ratelimit(long user, long server, NexusCommandCore command) {
        Entity key = new Entity(command.uuid, user);

        if (!ratelimits.containsKey(key)) {
            ratelimits.put(key, new NexusRatelimitDataCore(user));
        }

        NexusRatelimitDataCore entity = (NexusRatelimitDataCore) ratelimits.get(key);
        if (entity.isRatelimitedOn(server)) {
            // This if-statement defines the section where the user
            // is rate-limited.
            if (getRemainingSecondsFrom(server, entity, command) > 0) {
                if (!entity.isNotifiedOn(server)) {
                    entity.notified(server);
                    return new AccessorRatelimitData(false, true, getRemainingSecondsFrom(server, entity, command));
                }

                return new AccessorRatelimitData(true, true, getRemainingSecondsFrom(server, entity, command));
            }

            // This statement means that the user is still regarded as rate-limited by the
            // rate-limiter despite the cooldown expiring.
            entity.release(server);
        }

        entity.ratelimit(server);
        return new AccessorRatelimitData(false, false, -1);
    }

    /**
     * Gets the remaining seconds from the data provided.
     *
     * @param server The ID of the server.
     * @param dataCore The rate-limit core data.
     * @param commandCore The Nexus Command Core data.
     * @return The remaining seconds before the rate-limit should be released.
     */
    private long getRemainingSecondsFrom(long server, NexusRatelimitDataCore dataCore, NexusCommandCore commandCore) {
        return (TimeUnit.MILLISECONDS.toSeconds(
                dataCore.getRatelimitedTimestampInMillisOn(server)
                        +
                        commandCore.cooldown.toMillis()
        )) - TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis());
    }

    @Override
    public Optional<NexusRatelimitData> get(NexusCommand command, long user) {
        return Optional.ofNullable(ratelimits.getOrDefault(new Entity(((NexusCommandCore) command).uuid, user), null));
    }

    @Override
    public void onBeforeCommand(NexusMiddlewareEvent event) {
        AccessorRatelimitData ratelimitData = ratelimit(event.getUserId(), event.getServerId().orElse(event.getUserId()),
                (NexusCommandCore) event.getCommand());

        if (ratelimitData.ratelimited()) {
            if (!ratelimitData.notified()) {
                event.stop
                        (((NexusCore) event.getNexus())
                                .getMessageConfiguration().onRatelimited(event, ratelimitData.remaining())
                );
                return;
            }

            event.stop();
        }
    }

    private record AccessorRatelimitData(boolean notified, boolean ratelimited, long remaining) {
    }

    private record Entity(String commandUUID, long user) {

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            Entity entity = (Entity) o;
            return user == entity.user && Objects.equals(commandUUID, entity.commandUUID);
        }

        @Override
        public int hashCode() {
            return Objects.hash(commandUUID, user);
        }

    }

}

File: src/main/java/pw/mihou/nexus/features/command/interceptors/commons/modules/ratelimiter/core/NexusRatelimiterCore.kt
package pw.mihou.nexus.features.command.interceptors.commons.modules.ratelimiter.core

import pw.mihou.nexus.Nexus
import pw.mihou.nexus.features.command.core.NexusCommandCore
import pw.mihou.nexus.features.command.facade.NexusCommand
import pw.mihou.nexus.features.command.facade.NexusMiddlewareEvent
import pw.mihou.nexus.features.command.interceptors.commons.modules.ratelimiter.facade.NexusRatelimitData
import pw.mihou.nexus.features.command.interceptors.commons.modules.ratelimiter.facade.NexusRatelimiter
import java.time.Duration
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit
import kotlin.jvm.optionals.getOrDefault

internal class NexusRatelimiterCore : NexusRatelimiter {

    private val ratelimits: MutableMap<Entity, NexusRatelimitData?> = ConcurrentHashMap()

    /**
     * Ratelimits a user on a specific server, or on their private channel for a
     * specific command.
     *
     * @param user      The user to rate-limit.
     * @param server    The server or private channel to rate-limit on.
     * @param command   The command to rate-limit on.
     * @return          The results from the rate-limit attempt.
     */
    private fun ratelimit(user: Long, server: Long, command: NexusCommandCore): AccessorRatelimitData {
        val key = Entity(command.uuid, user)

        if (!ratelimits.containsKey(key)) {
            ratelimits[key] = NexusRatelimitDataCore(user)
        }

        val entity = ratelimits[key] as NexusRatelimitDataCore?
        if (entity!!.isRatelimitedOn(server)) {
            if (getRemainingSecondsFrom(server, entity, command) > 0) {
                if (!entity.isNotifiedOn(server)) {
                    entity.notified(server)
                    return AccessorRatelimitData(
                        notified = false,
                        ratelimited = true,
                        remaining = getRemainingSecondsFrom(server, entity, command)
                    )
                }
                return AccessorRatelimitData(
                    notified = true,
                    ratelimited = true,
                    remaining = getRemainingSecondsFrom(server, entity, command)
                )
            }

            entity.release(server)
        }
        entity.ratelimit(server)
        return AccessorRatelimitData(
            notified = false,
            ratelimited = false,
            remaining = -1
        )
    }

    /**
     * Gets the remaining seconds from the data provided.
     *
     * @param server The ID of the server.
     * @param dataCore The rate-limit core data.
     * @param commandCore The Nexus Command Core data.
     * @return The remaining seconds before the rate-limit should be released.
     */
    private fun getRemainingSecondsFrom(
        server: Long,
        dataCore: NexusRatelimitDataCore?,
        commandCore: NexusCommandCore
    ): Long {
        return TimeUnit.MILLISECONDS.toSeconds(
            dataCore!!.getRatelimitedTimestampInMillisOn(server)
                    +
                    commandCore.get("cooldown", Duration::class.java).getOrDefault(Duration.ofSeconds(5)).toMillis()
        ) - TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis())
    }

    override fun get(command: NexusCommand, user: Long): Optional<NexusRatelimitData> {
        return Optional.ofNullable(ratelimits.getOrDefault(Entity((command as NexusCommandCore).uuid, user), null))
    }

    override fun onBeforeCommand(event: NexusMiddlewareEvent) {
        val ratelimitData = ratelimit(
            event.userId, event.serverId.orElse(event.userId),
            event.command as NexusCommandCore
        )
        if (ratelimitData.ratelimited) {
            if (!ratelimitData.notified) {
                event.stop(
                    Nexus.configuration.commonsInterceptors.messages.ratelimited(event, ratelimitData.remaining)
                )
                return
            }
            event.stop()
        }
    }

    private inner class AccessorRatelimitData(val notified: Boolean, val ratelimited: Boolean, val remaining: Long)
    private inner class Entity(val command: String, val user: Long) {

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other == null || javaClass != other.javaClass) return false
            val entity = other as Entity
            return user == entity.user && command == entity.command
        }

        override fun hashCode(): Int {
            return Objects.hash(command, user)
        }
    }
}